### PR TITLE
Fix code for PHP 7.2

### PR DESCRIPTION
**Pull request by ozh**

### CODE REVIEW

**Comment by ozh** [comment]
WIP -- first pass. From [WP](https://github.com/WordPress/wordpress-develop/commit/f3a52234ea62de3a8fe0fbafa4a723b55041fbf2)

**Comment by ozh** [comment]
Success -- merge + tag new release ?

File: src/Parser/Plural.php [hunk 1]
```diff
@@ -0,0 +1,336 @@
+<?php
+
+/**
+ * A gettext Plural-Forms parser.
+ */
```
Comment (LeoColomb) [inline src/Parser/Plural.php]: This is the class description, is it possible to let it just above class definition?

File: src/Parser/Plural.php [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+    protected $cache = array();
+
+    /**
+     * Constructor.
+     *
+     * @param string $str Plural function (just the bit after `plural=` from Plural-Forms)
+     */
+    public function __construct($str)
+    {
+        $this->parse($str);
```
[the remark this answers is not in the document]
Comment (ozh) [reply src/Parser/Plural.php]: To make it obvious what the class does: when instantiated, it "parses". Putting whole this insane code in the constructor would leave you wondering "what the f*** does this class do??"

[the remark this answers is not in the document]
Comment (LeoColomb) [reply src/Parser/Plural.php]: Hé hé 😆 
But I disagree. [what the f*** does this class do?] is supposed to be provided both by the class name (`Plural` in `Parser` so it should parse something about plural) and the description/PHPDoc (that's why https://github.com/LeoColomb/pomo/pull/7#discussion_r142872216).
And actually, this hide how big and complicated is the parse function, which should be divided IMHO.

[the remark this answers is not in the document]
Comment (ozh) [reply src/Parser/Plural.php]: Absolutely, should be certainly divided in smaller parts, hence my comment about "this insane code" :)

[the remark this answers is not in the document]
Comment (LeoColomb) [reply src/Parser/Plural.php]: 😉 👍

**Comment by LeoColomb** [comment]
OK, not good enough for me (haha, yes, perfectionist), let me merge your work correctly. 😄

**Comment by ozh** [comment]
The last test in the class fails. No idea why. The concept it's supposed to test is fairly simple, so I'd say either fix it if you know why it fails, or just remove it.

**Comment by ozh** [comment]
Yep, spent another 40 mn on this mock test, still no idea why it fails. I suggest you get rid of it (unless of course you have an idea)

**Comment by LeoColomb** [comment]
No no, it's OK for me, everything is passing. Thanks, I will merge manually with my additions. 🎉 

---
Une question par contre, je suis pas certain de tout suivre. Regardons cette nouvelle fonction.

https://github.com/WordPress/wordpress-develop/blob/b7eac77fde8287945576339a3e0fe98ce3a10cb9/src/wp-includes/pomo/translations.php#L198-L212

On est d'accord qu'en cas de fail à l'intérieur du try, on passe à un appel de 
```php
return $this->make_plural_form_function(2, 'n != 1');
```
qui, si je ne m'abuse, appel la fonction dans laquelle on est (en mode recursif donc). 
Comment ça se termine ?

**Comment by LeoColomb** [comment]
Oh and another question, why did you split `test_exceptions()` function in unit tests?

**Comment by ozh** [comment]
Aucune idée, j'ai pas lu le code de toute cette PR. Confiance aveugle à l'auteur rmccue. J'en déduis simplement que la methode `Parser::parse()` ne **peut pas** foirer si l'argument est `'n != 1'`.

**Comment by ozh** [comment]
Regarding splitting `test_exceptions()` : I just started from their [initial commit](https://github.com/WordPress/wordpress-develop/commit/f3a52234ea62de3a8fe0fbafa4a723b55041fbf2#diff-fe6c7a0588f1f430884f16984e1e36a8R159) which is using one test per exception. I've seen the later commit, but I thought using `try catch` instead of the built-in annotation `@expectedException` was a bit lame.

**Pull request closed without merging**